OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs

```diff
@@ -99,53 +105,68 @@ public IngameChatLogic(Widget widget, OrderManager orderManager, World world, Mo
 					return true;
 			};
 
-			chatText.OnEscKey = () => { CloseChat(); return true; };
+			chatText.OnEscKey = () =>
+			{
+				if (!isMenuChat)
+					CloseChat();
+				else
+					chatText.YieldKeyboardFocus();
 
-			var chatClose = chatChrome.Get<ButtonWidget>("CHAT_CLOSE");
-			chatClose.OnClick += CloseChat;
+				return true;
+			};
 
-			chatPanel.OnKeyPress = e =>
+			if (!isMenuChat)
 			{
-				if (e.Event == KeyInputEvent.Up)
-					return false;
+				var chatClose = chatChrome.Get<ButtonWidget>("CHAT_CLOSE");
+				chatClose.OnClick += CloseChat;
 
-				if (!chatChrome.IsVisible() && (e.Key == Keycode.RETURN || e.Key == Keycode.KP_ENTER))
+				chatPanel.OnKeyPress = e =>
 				{
-					OpenChat();
-					return true;
-				}
+					if (e.Event == KeyInputEvent.Up)
+						return false;
 
-				return false;
-			};
+					if (!chatChrome.IsVisible() && (e.Key == Keycode.RETURN || e.Key == Keycode.KP_ENTER))
+					{
+						OpenChat();
+						return true;
+					}
+
+					return false;
+				};
+			}
 
 			chatScrollPanel = chatChrome.Get<ScrollPanelWidget>("CHAT_SCROLLPANEL");
 			chatTemplate = chatScrollPanel.Get<ContainerWidget>("CHAT_TEMPLATE");
 			chatScrollPanel.RemoveChildren();
 			chatScrollPanel.ScrollToBottom();
 
-			foreach (var chatLine in orderManager.ChatCache)
+			foreach (var chatLine in orderManager.ChatCache.Skip(isMenuChat && orderManager.ChatCache.Count() > 100 ? orderManager.ChatCache.Count() - 20 : 0))
```

Comment:
Yes, but ctor (that foreach) runs once for ingame chat, so it is done only first time (usually history small at this moment) while ctor for Chat tab runs each time you go to Options (or only when Chat tab, I don't remember, anymore more times).